File: backend/app/models/payment.py
from enum import Enum
from datetime import datetime
from sqlalchemy import Integer, String, ForeignKey, DateTime, Enum as SQLAlchemyEnum, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from backend.app.db.base_class import Base

class TransactionType(str, Enum):
    DEPOSIT = "DEPOSIT"                     # Money in -> Credits
    WITHDRAWAL = "WITHDRAWAL"               # Credits -> Money out (UC-19)
    BOOKING_HOLD = "BOOKING_HOLD"           # Credits held for booking
    BOOKING_RELEASE = "BOOKING_RELEASE"     # Credits paid to expert
    BOOKING_REFUND = "BOOKING_REFUND"       # Credits returned to student
    SERVICE_PAYMENT = "SERVICE_PAYMENT"     # Pay for AI service
    REFUND_REQUEST = "REFUND_REQUEST"       # UC-36: User requests bank withdrawal

class TransactionStatus(str, Enum):
    PENDING = "PENDING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    REFUNDED = "REFUNDED"                   # UC-36: Bank refund paid out by admin
    PENDING_PAYOUT = "PENDING_PAYOUT"       # UC-19: Expert withdrawal awaiting admin payout
    REJECTED_PAYOUT = "REJECTED_PAYOUT"    # UC-19: Admin rejected → credits refunded to expert

class PaymentTransaction(Base):
    __tablename__ = "transactions"
    
    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("user.id"), nullable=False)
    booking_id: Mapped[int] = mapped_column(Integer, ForeignKey("bookings.id"), nullable=True)
    
    amount: Mapped[int] = mapped_column(Integer, nullable=False) # Credits
    type: Mapped[TransactionType] = mapped_column(SQLAlchemyEnum(TransactionType), nullable=False)
    status: Mapped[TransactionStatus] = mapped_column(SQLAlchemyEnum(TransactionStatus), default=TransactionStatus.PENDING)
    
    description: Mapped[str] = mapped_column(String, nullable=True)
    payment_gateway_id: Mapped[str] = mapped_column(String, nullable=True) # For external tracking
    
    # UC-19: Snapshot of bank info at time of withdrawal request
    bank_name: Mapped[str] = mapped_column(String, nullable=True)
    bank_account: Mapped[str] = mapped_column(String, nullable=True)
    bank_holder_name: Mapped[str] = mapped_column(String, nullable=True)
    
    # UC-19: Admin verdict note when approving/rejecting
    admin_note: Mapped[str] = mapped_column(Text, nullable=True)
    
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    user = relationship("User", backref="transactions")
    booking = relationship("Booking", backref="transactions")

File: frontend/src/types/payment.ts
export enum TransactionType {
    DEPOSIT = "DEPOSIT",
    WITHDRAWAL = "WITHDRAWAL",
    BOOKING_HOLD = "BOOKING_HOLD",
    BOOKING_RELEASE = "BOOKING_RELEASE",
    BOOKING_REFUND = "BOOKING_REFUND",
    SERVICE_PAYMENT = "SERVICE_PAYMENT",
    REFUND_REQUEST = "REFUND_REQUEST",   // UC-36
}

export enum TransactionStatus {
    PENDING = "PENDING",
    COMPLETED = "COMPLETED",
    FAILED = "FAILED",
    REFUNDED = "REFUNDED",              // UC-36: bank transfer done
    PENDING_PAYOUT = "PENDING_PAYOUT",  // UC-19: awaiting admin payout
    REJECTED_PAYOUT = "REJECTED_PAYOUT", // UC-19: admin rejected
}

export interface PaymentTransaction {
    id: number;
    amount: number;
    type: TransactionType;
    status: TransactionStatus;
    created_at: string;
    description?: string;
    bank_name?: string;
    bank_account?: string;
    bank_holder_name?: string;
    admin_note?: string;
    payout_reference?: string;
    user?: {
        id: number;
        email: string;
        full_name?: string;
        credits: number;
        expert_profile?: {
            id: number;
            kyc_status: string;
        };
    };
}

export interface PaymentCreate {
    amount: number;
}

export interface PaymentDepositResponse {
    transaction_id: number;
    qr_url: string;
    amount_vnd: number;
    content: string;
}

export interface PaginatedPaymentResponse {
    items: PaymentTransaction[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
}

// UC-36
export interface RefundRequestCreate {
    amount: number;
    bank_name: string;
    bank_account: string;
    account_holder: string;
}

// UC-19: Expert Withdrawal
export interface WithdrawalCreate {
    amount: number;
}

export interface WithdrawalRequest extends PaymentTransaction {
    user_id: number;
}

export interface ExpertStats {
    available: number;
    escrow: number;
    monthly_total: number;
    trend: number;
}
